File: 알고리즘 스터디 문제들/Part 7 - 백트래킹/추가 문제/부분수열의 합.cpp
#include <iostream>
#include <array>
using namespace std;

int N, S;
int cnt = 0;
array<int, 20> arr;

void backtrack(int idx, int total)
{
    if (idx == N)
    {
        if (total == S)
            cnt++;
        return;
    }

    backtrack(idx + 1, total);
    backtrack(idx + 1, total + arr[idx]);
}

int main()
{
    cin >> N >> S;
    for (size_t i = 0; i < N; i++)
        cin >> arr[i];

    backtrack(0, 0);
    cout << ((S == 0) ? cnt - 1 : cnt);
}

File: 알고리즘 스터디 문제들/Part 7 - 백트래킹/추가 문제/부분수열의 합.py
def backtrack(cur, total, nums, N, S):
    if (cur == N):
        if (total == S):
            return 1
        else:
            return 0

    cnt = 0
    cnt += backtrack(cur + 1, total, nums, N, S)
    cnt += backtrack(cur + 1, total + nums[cur], nums, N, S)
    return cnt


N, S = map(int, input().split())
nums = list(map(int, input().split()))

if (S == 0):
    print(backtrack(0, 0, nums, N, S) - 1)
else:
    print(backtrack(0, 0, nums, N, S))